Documented exception thrown is incorrect for javax.xml.bind.DatatypeConverter#print(Object) methods

Previously tracked via: https://bugs.openjdk.java.net/browse/JDK-8152531

A DESCRIPTION OF THE PROBLEM : 
The documentation for the javax.xml.bind.DatatypeConverter#print(Object) methods (such as printDate, printDateTime and printDecimal) state an IllegalArgumentException is thrown if the input parameter is null. In fact, a NullPointerException is thrown as shown in the Junit sample below: 

    @Test 
    public void testJDate() { 

        this.exception.expect(IllegalArgumentException.class); 
        javax.xml.bind.DatatypeConverter.printDecimal(null); 

    } 


java.lang.AssertionError: 
Expected: an instance of java.lang.IllegalArgumentException 
     but: <java.lang.NullPointerException> is a java.lang.NullPointerException 
Stacktrace was: java.lang.NullPointerException 
at javax.xml.bind.DatatypeConverterImpl._printDecimal(DatatypeConverterImpl.java:546) 
at javax.xml.bind.DatatypeConverterImpl.printDecimal(DatatypeConverterImpl.java:542) 
at javax.xml.bind.DatatypeConverter.printDecimal(DatatypeConverter.java:514) 
        ... 


Of course, if it really SHOULD be an IllegalArgumentException, someone might want to open a jre bug report. 

There is minimal impact as it is now unclear what should be thrown by a custom converter method. 


URL OF FAULTY DOCUMENTATION : 
http://docs.oracle.com/javase/8/docs/api/javax/xml/bind/DatatypeConverter.html

--- a/api/src/main/java/jakarta/xml/bind/DatatypeConverterImpl.java
+++ b/api/src/main/java/jakarta/xml/bind/DatatypeConverterImpl.java
@@ -66,6 +66,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
     }
 
     public static String _printInteger(BigInteger val) {
+        if (null == val) throw new IllegalArgumentException("val is null");
         return val.toString();
     }
 
@@ -423,6 +424,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
     }
 
     public static String _printDateTime(Calendar val) {
+        if (null == val) throw new IllegalArgumentException("val is null");
         return CalendarFormatter.doFormat("%Y-%M-%DT%h:%m:%s%z", val);
     }
 
@@ -471,6 +473,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
 
     @Override
     public String printHexBinary(byte[] data) {
+        if (null == data) throw new IllegalArgumentException("data is null");
         StringBuilder r = new StringBuilder(data.length * 2);
         for (byte b : data) {
             r.append(hexCode[(b >> 4) & 0xF]);
@@ -501,6 +504,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
 
     @Override
     public String printTime(Calendar val) {
+        if (null == val) throw new IllegalArgumentException("val is null");
         return CalendarFormatter.doFormat("%h:%m:%s%z", val);
     }
 
@@ -515,6 +519,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
     }
 
     public static String _printDate(Calendar val) {
+        if (null == val) throw new IllegalArgumentException("val is null");
         return CalendarFormatter.doFormat((new StringBuilder("%Y-%M-%D").append("%z")).toString(),val);
     }
 
@@ -554,6 +559,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
     }
 
     public static String _printDecimal(BigDecimal val) {
+        if (null == val) throw new IllegalArgumentException("val is null");
         return val.toPlainString();
     }
 
@@ -776,6 +782,7 @@ final class DatatypeConverterImpl implements DatatypeConverterInterface {
     }
 
     public static String _printBase64Binary(byte[] input) {
+        if (null == input) throw new IllegalArgumentException("input is null");
         return _printBase64Binary(input, 0, input.length);
     }
 

--- a/api/src/main/java/jakarta/xml/bind/DatatypeConverterImpl.java
+++ b/api/src/main/java/jakarta/xml/bind/DatatypeConverterImpl.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2007, 2021 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2007, 2023 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Distribution License v. 1.0, which is available at

--- a/api/src/test/java/org/eclipse/jaxb/api/DatatypeConverterTest.java
+++ b/api/src/test/java/org/eclipse/jaxb/api/DatatypeConverterTest.java
@@ -55,4 +55,30 @@ public class DatatypeConverterTest {
         Assert.assertEquals(true, DatatypeConverter.parseBoolean("true "));
         Assert.assertEquals(true, DatatypeConverter.parseBoolean(" true "));
     }
+
+    @Test 
+    public void testJDate() { 
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printInteger(null));
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printDateTime(null));
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printHexBinary(null));
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printTime(null));
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printDate(null));
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printDecimal(null));
+        Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printBase64Binary(null));
+        
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printShort(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printFloat(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printBoolean(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printByte(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printUnsignedInt(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printString(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printInt(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printLong(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printDouble(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printQName(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printUnsignedShort(null));
+        //Assert.assertThrows(IllegalArgumentException.class, () -> DatatypeConverter.printAnySimpleType(null));
+        
+    } 
 }
+

